File: tests/algorithms/x/Kotlin/maths/euler_modified.kt
var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun ceil_float(x: Double): Int {
    var i: Int = ((x.toInt())).toInt()
    if (x > ((i.toDouble()))) {
        return i + 1
    }
    return i
}

fun exp_approx(x: Double): Double {
    var term: Double = 1.0
    var sum: Double = 1.0
    var n: Int = (1).toInt()
    while (n < 20) {
        term = (term * x) / ((n.toDouble()))
        sum = sum + term
        n = n + 1
    }
    return sum
}

fun euler_modified(ode_func: (Double, Double) -> Double, y0: Double, x0: Double, step: Double, x_end: Double): MutableList<Double> {
    var n: Int = (ceil_float((x_end - x0) / step)).toInt()
    var y: MutableList<Double> = mutableListOf(y0)
    var x: Double = x0
    var k: Int = (0).toInt()
    while (k < n) {
        var y_predict: Double = y[k]!! + (step * ode_func(x, y[k]!!))
        var slope1: Double = ((ode_func(x, y[k]!!)).toDouble())
        var slope2: Double = ((ode_func(x + step, y_predict)).toDouble())
        var y_next: Double = y[k]!! + ((step / 2.0) * (slope1 + slope2))
        y = run { val _tmp = y.toMutableList(); _tmp.add(y_next); _tmp }
        x = x + step
        k = k + 1
    }
    return y
}

fun f1(x: Double, y: Double): Double {
    return (((0.0 - 2.0) * x) * y) * y
}

fun f2(x: Double, y: Double): Double {
    return ((0.0 - 2.0) * y) + (((x * x) * x) * exp_approx((0.0 - 2.0) * x))
}

fun user_main(): Unit {
    var y1: MutableList<Double> = euler_modified(::f1, 1.0, 0.0, 0.2, 1.0)
    println(y1[y1.size - 1]!!)
    var y2: MutableList<Double> = euler_modified(::f2, 1.0, 0.0, 0.1, 0.3)
    println(y2[y2.size - 1]!!)
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        user_main()
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}

File: tests/algorithms/x/PHP/maths/euler_modified.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _append($arr, $x) {
    $arr[] = $x;
    return $arr;
}
$__start_mem = memory_get_usage();
$__start = _now();
  function ceil_float($x) {
  $i = intval($x);
  if ($x > (floatval($i))) {
  return $i + 1;
}
  return $i;
};
  function exp_approx($x) {
  $term = 1.0;
  $sum = 1.0;
  $n = 1;
  while ($n < 20) {
  $term = $term * $x / (floatval($n));
  $sum = $sum + $term;
  $n = $n + 1;
};
  return $sum;
};
  function euler_modified($ode_func, $y0, $x0, $step, $x_end) {
  $n = ceil_float(($x_end - $x0) / $step);
  $y = [$y0];
  $x = $x0;
  $k = 0;
  while ($k < $n) {
  $y_predict = $y[$k] + $step * $ode_func($x, $y[$k]);
  $slope1 = $ode_func($x, $y[$k]);
  $slope2 = $ode_func($x + $step, $y_predict);
  $y_next = $y[$k] + ($step / 2.0) * ($slope1 + $slope2);
  $y = _append($y, $y_next);
  $x = $x + $step;
  $k = $k + 1;
};
  return $y;
};
  function f1($x, $y) {
  return -2.0 * $x * $y * $y;
};
  function f2($x, $y) {
  return -2.0 * $y + ($x * $x * $x) * exp_approx(-2.0 * $x);
};
  function main() {
  $y1 = euler_modified('f1', 1.0, 0.0, 0.2, 1.0);
  echo rtrim(json_encode($y1[count($y1) - 1], 1344)), PHP_EOL;
  $y2 = euler_modified('f2', 1.0, 0.0, 0.1, 0.3);
  echo rtrim(json_encode($y2[count($y2) - 1], 1344)), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;
